### PR TITLE
refactor(core): improvements to Construct API

### LINKED ISSUE #1934: Cleanup/revisit Construct API
Take another look at the APIs provided by the low-level `Construct` (and `ConstructNode`) classes and make sure we clean them up as much as possible, and reduce the surface area to the max.

### LINKED ISSUE #2766: Deprecate `node.stack`
Because it's a leak? And also, `Stack.of(this).boom` is shorter than `this.node.stack.boom`.

### PR DESCRIPTION
**Pull request by eladb**
Closes #1934

BREAKING CHANGE:
* **core:** `node.stack` is now `Stack.of(construct)` (fixes #2766)
* **core:** `node.resolve` has been moved to `stack.resolve`.
* **core:** `node.stringifyJson` has been moved to `stack.stringifyJson`.
* **core:** `node.validateTree` is now `ConstructNode.validate(node)`
* **core:** `node.prepareTree` is now `ConstructNode.prepare(node)`
* **core:** `node.getContext` is now `node.tryGetContext`
* **core:** `node.recordReference` is now `node.addReference`
* **core:** `node.apply` is now `node.applyAspect`
* **core:** `node.ancestors()` is now `node.scopes`
* **core:** `node.required` has been removed.
* **core:** `node.typename` has been removed.
* **core:** `node.addChild` is now private
* **core:** `node.findReferences()` is now `node.references`
* **core:** `node.findDependencies()` is now `node.dependencies`
* **core:** `stack.dependencies()` is now `stack.dependencies`
* **core:** `CfnElement.stackPath` has been removed.
* **core:** `CloudFormationLang` is now internal (use `stack.toJsonString()`)

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

**Comment by rix0rrr** [comment]
I think the last commit does things that aren't necessary:
> `node.resolve` has been moved to `stack.resolve`. 

Weren't we talking about removing `resolve()` altogether? Better to do that than to move it then.

> `node.stringifyJson` has been moved to `stack.stringifyJson`.

Actually this method could go as well. It does nothing that `CloudFormationLang.toJSON()` doesn't do, so the usage sites could be replaced.

Yes, this couples it to CloudFormation (we lose target language abstraction), but in all instances where we're calling it, we're calling it from the context of a Construct Library resource which already knows it's going to be deployed using CloudFormation (same situation as where we would be using `Stack.of()`) so the coupling is justified.

**Comment by eladb** [comment]
> Weren't we talking about removing resolve() altogether? Better to do that than to move it then.

I tried to look into moving `resolve()` into the synthesis session, and realized it's going to be too hard. It is needed in some "prepare" contexts as well and an infinite number of tests. I was concerned that it was under `node`, but under `stack` it feels better, especially given it uses CloudFormation functions for concatenations, etc. So I am comfortable with `stack.resolve` for now. If you feel strongly about it, I am willing to give this another try.

> Actually this method could go as well. It does nothing that CloudFormationLang.toJSON() doesn't do, so the usage sites could be replaced.

I think `stack.stringifyJson` is more discoverable and easy to reason about than `CloudFormationLang.toJSON()` which we should make internal. And it is used quite a lot (every time there's a "document" resource property), and I believe customers will potentially use it with "leaky" custom resources as well.

File: packages/@aws-cdk/cdk/lib/stack.ts [hunk 1]
```diff
@@ -179,41 +181,21 @@ export class Stack extends Construct implements ITaggable {
   }
 
   /**
-   * Returns the CloudFormation template for this stack by traversing
-   * the tree and invoking _toCloudFormation() on all Entity objects.
-   *
-   * @internal
+   * Resolve a tokenized value in the context of the current Construct
```
Comment (RomainMuller) [inline packages/@aws-cdk/cdk/lib/stack.ts]: Well - in the context of the current `Stack` then ¯\\\_(シ)\_/¯

File: packages/@aws-cdk/cdk/lib/stack.ts [hunk 2]
```diff
@@ -179,41 +181,21 @@ export class Stack extends Construct implements ITaggable {
...
 
-      return ret;
-    } finally {
-      // allow mutations after synthesis is finished.
-      this.node.unlock();
-    }
+  /**
+   * Convert an object, potentially containing tokens, to a JSON string
+   */
+  public stringifyJson(obj: any): string {
```
Comment (RomainMuller) [inline packages/@aws-cdk/cdk/lib/stack.ts]: This does not *stringify JSON*, it *stringifies stuff to CFN-y Json*. I don't love that method name, and I kinda think `stringify` or `toJsonString` is better.

Comment (eladb) [reply packages/@aws-cdk/cdk/lib/stack.ts]: You are right. I like `toJsonString`.

Comment (eladb) [reply packages/@aws-cdk/cdk/lib/stack.ts]: @rix0rrr what do you think?

Comment (rix0rrr) [reply packages/@aws-cdk/cdk/lib/stack.ts]: Yeah cool

**Pull request merged**